GitHubCodespaces.ps1

```diff
@@ -0,0 +1,538 @@
+    param(
+        [Parameter(Mandatory,
+            ValueFromPipeline,
+            ValueFromPipelineByPropertyName)]
+        [string] $CodespaceName,
+
+        [switch] $PassThru,
+
+        [switch] $Wait,
+
+        [string] $AccessToken
+    )
+
+    process
+    {
+        Write-InvocationLog
+
+        $telemetryProperties = @{
+            'CodespaceName' = Get-PiiSafeString -PlainText $CodespaceName
+        }
+
+        $params = @{
+            'UriFragment' = "user/codespaces/$CodespaceName/stop"
+            'Method' = 'POST'
+            'Description' = "Stop Codespace $CodespaceName"
+            'AccessToken' = $AccessToken
+            'AcceptHeader' = $script:symmetraAcceptHeader
+            'TelemetryEventName' = $MyInvocation.MyCommand.Name
+            'TelemetryProperties' = $telemetryProperties
+        }
+
+        if (-not $PSCmdlet.ShouldProcess($CodespaceName, "Stop Codespace $CodespaceName"))
+        {
+            return
+        }
+
+        $result = (Invoke-GHRestMethod @params | Add-GitHubCodespaceAdditionalProperties)
+
+        if ($Wait.IsPresent)
+        {
+            $waitParams = @{
+                'CodespaceName' = $CodespaceName
+                'AccessToken' = $AccessToken
+            }
+
+            $result = Wait-GitHubCodespaceAction @waitParams
+        }
+
+        if (Resolve-ParameterWithDefaultConfigurationValue -Name PassThru -ConfigValueName DefaultPassThru)
+        {
+            return $result
+        }
+    }
+}
+
+function Wait-GitHubCodespaceAction
+{
+    <#
+    .SYNOPSIS
+        Wait for a Codespace start or stop action.
+
+    .PARAMETER CodespaceName
+        The name of the codespace.
+
+    .PARAMETER AccessToken
+        If provided, this will be used as the AccessToken for authentication with the
+        REST Api.  Otherwise, will attempt to use the configured value or will run unauthenticated.
+
+    .INPUTS
+        GitHub.Codespace
+
+    .OUTPUTS
+        GitHub.Codespace
+
+    .EXAMPLE
+        Wait-GitHubCodespace -Name vercellone-effective-goggles-qrv997q6j9929jx8
+#>
+    [CmdletBinding()]
+    [OutputType({$script:GitHubCodespaceTypeName})]
+    param(
+        [Parameter(Mandatory,
+            ValueFromPipeline,
+            ValueFromPipelineByPropertyName)]
+        [string] $CodespaceName,
+
+        [string] $AccessToken
+    )
+
+    begin {
+        # 2s minimum
+        $sleepSeconds = $(Get-GitHubConfiguration -Name 'StateChangeDelaySeconds')
+        if ($sleepSeconds -eq 0)
+        {
+            $sleepSeconds = 2
+        }
+    }
+
+    process
+    {
+        Write-InvocationLog
+
+        do
+        {
+            Start-Sleep -Seconds $sleepSeconds
+            $codespace = (Get-GitHubCodespace @PSBoundParameters)
+        } until($codespace.state -notmatch 'Queued|ing')
```

Comment:
This is a bit tricky.  I understand what you're doing, but can you document in a comment what the possible '*ing' states might be?